kaiax/valset/impl/getter_context.go

```diff
@@ -0,0 +1,275 @@
+// Copyright 2024 The Kaia Authors
+// This file is part of the Kaia library.
+//
+// The Kaia library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The Kaia library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the Kaia library. If not, see <http://www.gnu.org/licenses/>.
+
+package impl
+
+import (
+	"math"
+	"math/big"
+
+	"github.com/kaiachain/kaia/blockchain/types"
+	"github.com/kaiachain/kaia/common"
+	"github.com/kaiachain/kaia/consensus/istanbul"
+	"github.com/kaiachain/kaia/kaiax/gov"
+	"github.com/kaiachain/kaia/kaiax/staking"
+	"github.com/kaiachain/kaia/kaiax/valset"
+	"github.com/kaiachain/kaia/params"
+)
+
+// blockContext is the common input for committee and proposer selection for a block.
+// Because committee and proposer selection calls each other, we collect common inputs here.
+type blockContext struct {
+	num uint64
+
+	// Computed inputs
+	qualified    *valset.AddressSet
+	prevHeader   *types.Header
+	prevProposer common.Address
+	// convenience fields
+	rules params.Rules
+	pset  gov.ParamSet
+}
+
+func (v *ValsetModule) getBlockContext(num uint64) (*blockContext, error) {
+	qualified, err := v.getQualifiedValidators(num)
+	if err != nil {
+		return nil, err
+	}
+	prevHeader := v.Chain.GetHeaderByNumber(num - 1)
+	if prevHeader == nil {
+		return nil, errNoHeader
+	}
+	prevProposer, err := v.Chain.Engine().Author(prevHeader)
+	if err != nil {
+		return nil, err
+	}
+
+	return &blockContext{
+		num:          num,
+		qualified:    qualified,
+		prevHeader:   prevHeader,
+		prevProposer: prevProposer,
+		rules:        v.Chain.Config().Rules(new(big.Int).SetUint64(num)),
+		pset:         v.GovModule.EffectiveParamSet(num),
+	}, nil
+}
+
+func (v *ValsetModule) getCommittee(c *blockContext, round uint64) ([]common.Address, error) {
+	if c.num == 0 {
+		return c.qualified.List(), nil
+	}
+
+	currProposer, err := v.getProposer(c, round)
+	if err != nil {
+		return nil, err
+	}
+	nextDistinctProposer := currProposer // TOOD
```

Comment:
In case you forgot:

```suggestion
	nextDistinctProposer := currProposer // TODO: ~~
```

Reply:
I'll add v.getNextDistinctProposer and ping you.

Reply:
Done.